nonce/nonce.go

```diff
@@ -117,6 +121,11 @@ func (ns *NonceService) encrypt(counter int64) (string, error) {
 	ct := ns.gcm.Seal(nil, nonce, pt, nil)
 	copy(ret, nonce[4:])
 	copy(ret[8:], ct)
+
+	// If we are using a identifying prefix, append it to the nonce
+	if ns.prefix != nil {
+		ret = append([]byte{*ns.prefix}, ret...)
```

Comment:
Since we're prepending the prefix as a byte, then base64 encoding, the prefix of the base64-encoded version (what we'll see during integration tests etc) won't always match what's in the config file (hex-encoded).

I think it would be better, since we config the prefix as a hex string, to keep as hex all the way through. So that would imply encoding it as hex here, encoding the nonce body, and concatenating the two. I think that will make debugging a lot easier and less confusing.

Reply:
I'm not super into the idea of using mixed encodings in a string blob. As far as I can tell from a quick glance over logs and integration tests we don't log the nonce anywhere. I think in any circumstance where we'd be debugging this you'd already need to add extra code, and at that point you'd be able to just add code to parse out the first byte.

I'd be slightly less opposed to adding some structure to the nonce in order to differentiate the hex encoded portion from the base64 encoded portion, but even them I'm not really convinced the added complexity is worth the benefits it gives us.

Reply:
What about storing the prefix as base64 in the config files? If we use a three-character encoded prefix, it decodes to two bytes, and also will be straightforwardly concatenated onto the encoded nonce.

We don't currently log nonces, but we will know which instances are configured with which prefixes, and we can easily set `LOGLEVEL=0` on a chisel instance to see nonces being exchanged. In general I think this change will add enough complexity that we should prioritize straightforward debuggability.

Reply:
That'd work for me, although I believe we'd need to use a 4 character/3 byte prefix as base64 operates on 3 byte binary segments, with two bytes the first byte of the nonce (third byte overall) would alter the third character.

Reply:
Thinking about this slightly more, I guess the prefix doesn't need to be encoded _anything_, it just needs to be a string we match against. There is a tiny tiny chance some client somewhere has decided to decode the nonce and store it as bytes, which may incline us towards keeping it valid base64, but apart from that there is nothing stopping us from just doing something like `asd.n6ANQzxvZj6-KS9ygx1JrmRiDjhnC1jozNp9vspyHrQ` where `asd` would be the prefix. This would create a clear delimitation of what is the prefix and what is the nonce, doesn't impose weird size requirements, and makes parsing slightly simpler... what do you think?

Reply:
> There is a tiny tiny chance some client somewhere has decided to decode the nonce and store it as bytes

😞

https://github.com/certbot/certbot/blob/b0d960f102c998d8231c0ee48952b488f10864ac/acme/acme/client.py#L1148-L1158

Reply:
Back to 4 char base64 then :/

Reply:
Yeah also the spec says nonces have to be in the base64url alphabet.